### DISCUSSION TITLE
How can I properly add html in a json file?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#40369] [Help] [answered]
For example I have this line here:

```
[
{
"stuff here" : "more stuff and code below"
},
{
"description": "Here is a collection of hydraulic formulas <a href =\"https://www.tu.biz/fluid-power-formulas\"> click here to check it</a> ",
}
]
```

So, yes... I want to create a string under description, and then add an a tag with a link of where to go...   how can I do that?   obviously I am not using the escape quotes properly...

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Not really, it is ok. For completeness say you can use single quotes too:

```json
{
  "description": "Here is a collection of hydraulic formulas <a href='https://www.tu.biz/fluid-power-formulas'> click here to check it</a>"
}
```

The question is, how do we render this so that it doesn't do this:

![Screenshot 2022-09-09 at 08 52 48](https://user-images.githubusercontent.com/21013447/189289277-60e5fc18-4630-49d8-b67c-4120f0fc93d1.png)

And instead does this:

![Screenshot 2022-09-09 at 08 53 24](https://user-images.githubusercontent.com/21013447/189289371-94682e5e-4286-4a03-974e-c567e3e2264b.png)

The answer is something called, `dangerouslySetInnerHTML` and it is a lesser known React prop, which you can find in the React.js docs: https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml

> For completeness sake once again, you could parse the string with DOM APIs, and append that result to the DOM. This way you could also sanitize the HTML. This however is computationally expensive, and places DOM manipulation responsibility upon yourself. 
>  ```js
> const tagString = "Hi there: <div>I am a div node</div>";
> const fragment = document.createRange().createContextualFragment(tagString);
> document.body.appendChild(fragment)
>  ```
>
> And yet another path is to model your data (JSON) so that it is more like React's components, each child with a type and a children, which can in turn be components or raw strings. `"description": [{type:"string", children:["text"]}, {type: "a", children: [""], href=""}]` etc, etc... but that's more complex and not always needed.

In short, React.js takes care of escaping strings to prevent XSS, and various other issues. However, in some cases we'd like to render a string with HTML and that's when, as long as the string with HTML is not generated from user input, we can use `dangerouslySetInnerHTML`:

> Note: I modified your JSON structure so that you have do implement this yourself. It is important to be aware of what `dangerouslySetInnerHTML` is doing.

```js
const Home: NextPage = () => {
  return (
    <div>
      <div dangerouslySetInnerHTML={{ __html: data.description }}></div>
    </div>
  );
};
```

**Reply by icyJoseph** [reply]
> I would need to include the dangerouslySetInnerHTML at the component, then call the item.description at the page...

You just pass the data to the `dangerouslySetInnerHTML` as you normally do:

```jsx
const Foo = ({htmlText}) => {
  return (
      <div dangerouslySetInnerHTML={{ __html: htmlText }}></div>
  );
};
```

It isn't very special, aside from the fact that it might be dangerous! 😅 

> Most pages won't need html in the json.... wonder how the dangerouslySetInnerHTML behave if I don't have escaping quotes.... we will test it :)

If you don't properly escape the quotes in the JSON, then you'll have invalid JSON.

> Another solution would be to create a brand new component just for that one page and include the html there... but of course I would like to work it from the json and have less files...

Hardcoding the links you mean? Or perhaps have it all in a JSON file but structured:

```json
{
     "links": [{"href": "something", "text":"something-else"}]
}
```

At the end of the day it is a design choice, whether to pass things down as props or read it on place from a constant. @Elindo586

**Reply by icyJoseph** [reply]
@Elindo586:
 
```jsx
<div className="">
           <p dangerouslySetInnerHTML={{ __html:  }} />
</div>
```
 
That ought to work, but the problem is that `description` in this case, doesn't contain any HTML, it's a plain string if you will. 

> Understanding the difference between a fixed displacement pump vs. a variable displacement pump.

There's no HTML tags on that string.

I checked this page though: `pages/blog/english/fixed-displacement-pump-vs-variable-displacement-pump.js` maybe there's another example. 

Try to refer to the docs, they show how to use this correctly.

**Reply by Elindo586** [reply]
Yes, the majority of description items don't have html...  that is why I just wrote a sample on the first post...

Now... I think I understand how to get a variable and insert it the `dangerouslySetInnerHTML={{ __html:  }}`, this seems to be easier to do from a page.

But in this case I have a couple considerations to think about.

1. I am using a reusable component with props... and I don't know how to set the dangerouslySetInnerHTML as  a prop,  and later on take different values for the different pages that I am using the reusable component.

2.  Because not all description items need to have html, then I don't know how the dangerouslySetInnerHTML may render different type of information.

Reusable components are cool to work with when all the information has the same structure, but as soon as I need to change something then its a bit tricky.

**Reply by icyJoseph** [reply]
> 1. I am using a reusable component with props... and I don't know how to set the dangerouslySetInnerHTML as  a prop,  and later on take different values for the different pages that I am using the reusable component.

What do you mean exactly? You pass it as you pass any other prop.

> 2.  Because not all description items need to have html, then I don't know how the dangerouslySetInnerHTML may render different type of information.

I think as long as the data type is string you're good to go.

**Reply by icyJoseph** [reply]
I think anyway that the root issue has been solved, or?

**Reply by Elindo586** [reply]
I think I am missing some silly that I can't see... 

I go back to this props below.....   this is part of the component page code... 
```
<div className="">
         
            <p> {props.description}</p>
          </div>

```

Then in a separate page,  maps the json file, import the component into the page, and go like this...

```
<Blogger 2
description = {item.description} />
```

My head still doesn't know how to incorporate the `dangerouslySetInnerHTML={{ __html: }}` at the component page, so that later on  in a different page I can just do `description = {item.description}` and get the information I need on the page.

**Reply by icyJoseph** [reply]
Why would the page do that? It's only at the DOM element level that you invoke dangerously set inner HTML. Meaning that it only exists for say: `div`, `p`, etc, all native HTML tags. So you don't have to do anything special when your page component sends that in as props. Only directly at the `p` tag.

**Reply by Elindo586** [reply]
Ok so....  I don't know how to set it up at the component level file...

But calling component at the actual page file I can do some like this and it works:

```
<Blogger2 
description={
                <div
                  dangerouslySetInnerHTML={{ __html: item.description }}
                ></div>
              }
/>
```
^^^  Now that is cool... I can put it there instead of trying to put it at the component file...

Also.... as slow as I am... it made me think that I could just create a bunch of divs, p, a,  tags or whatever after the `description=`    If I want or need to setup something more elaborated I can work from there instead of the json file....   so I think?

Anyway... its working  .....   as always thank you much for your help and input... :)

**Comment by Elindo586** [comment]
OK... let me process it and test it some time today...

I would need to include the dangerouslySetInnerHTML at the component, then call the item.description at the page...

Most pages won't need html in the json.... wonder how the dangerouslySetInnerHTML behave if I don't have escaping quotes....  we will test it :)

Another solution would be to create a brand new component just for that one page and include the html there... but of course I would like to work it from the json and have less files...

**Comment by Elindo586** [comment]
Ok... how would you include the dangerouslySetInnerHTML  in the component props? 

```
<div className="">
           
            <p>{props.description}</p>
          </div>
```

^^^ this is what I have..  then at the page level I just map the json and get item.description and send it to fill  to the component spot...

I have been trying a number of combinations and haven't gotten it right ...

```
<div className="">
            {" "}
            <p> dangerouslySetInnerHTML={{ __html: {`${props.description}`} }}</p>
          </div>
```
the component is this one 
description on line 10
https://github.com/Elindo586/nextprojectb/blob/main/components/blog-post-english.js
